### PR TITLE
added support for bigger R(2+1)d fixes #7

### LINKED ISSUE #7: Use bigger R(2+1)d pre-trained on IG65M
Hey there, awesome project you have here :tada: 

I checked your code and seeing

https://github.com/v-iashin/video_features/blob/71a9a08d6a350a589fd275ff7f94803757573bca/models/r21d/extract_r21d.py#L9

I was wondering if you have thought about dropping this one in

https://github.com/moabitcoin/ig65m-pytorch

since the r(2+1)d 18 features are - based on experience - not the best to work with, in the r(2+1)d family.

:v:

### PR DESCRIPTION
**Pull request by ohjho**
added a new variable `model_name` for selecting other flavours of r21d. `r2plus1d_34_32_kinetics` and `r2plus1d_34_32_ig65m` made available

### CODE REVIEW

**Comment by v-iashin** [comment]
Hi @ohjho 

This is interesting and thanks for your input!

May I ask you for:
- [ ] to explain why do you think we should drop the step_size h-parameter?
- [ ] to implement the `show_preds` flag for `ig65m` and `kinetics`. 
  - [ ] It involves, finding label map (label list) similar to https://github.com/v-iashin/video_features/tree/master/utils. 
  - [ ] Some fixes should be made here: https://github.com/v-iashin/video_features/blob/master/models/r21d/extract_r21d.py#L113-L116
  - By using the model for classification, we (and end-users) could understand if the weights and implementation are correct.

**Comment by ohjho** [comment]
> 
> This is interesting and thanks for your input!

thank you! this is a great package, learning lots from it, and happy to help out wherever I can 

> 
> May I ask you for:
> 
>     * [ ]  to explain why do you think we should drop the step_size h-parameter?
no intention of dropping it, totally agree that `step_size` is a very important parameter, I was just a bit lazy with my loading of the defaults; will change it so that it will be more explicit/ pythonic

> 
>     * [ ]  to implement the `show_preds` flag for `ig65m` and `kinetics`.
>       
>       * [ ]  It involves, finding label map (label list) similar to https://github.com/v-iashin/video_features/tree/master/utils.
Great! thanks for the hints! I just thought the mapping was loaded automatically because I did see predicted labels using the `ig65m` models. This is a silly mistake.
>       * [ ]  Some fixes should be made here: https://github.com/v-iashin/video_features/blob/master/models/r21d/extract_r21d.py#L113-L116
I'll add a fix for this. Thanks for the detailed review and quick response :)
>       * By using the model for classification, we (and end-users) could understand if the weights and implementation are correct.

**Comment by ohjho** [comment]
hey @v-iashin ,

just updated `R21D_MODEL_LIST` with two new key-values `step_size` and `dataset`.

Seems like the ig65m model [does not come with a label map](https://github.com/moabitcoin/ig65m-pytorch#pytorch-and-onnx-models-trophy), so in `R21D_MODEL_LIST` I have only included the two versions of the 34-layer model that's fine-tuned on kinetics-400

File: configs/r21d.yml [hunk 1]
```diff
@@ -1,7 +1,8 @@
 # Model
 feature_type: 'r21d'
-stack_size: 16 # Feature time span in fps
-step_size: 16 # Feature step size in fps
+model_name: 'default'
```
Comment (v-iashin) [inline configs/r21d.yml:3]: Ok, but can we name it more informatively? E.g. `default` --> `r2plus1d_18_16_kinetics`

File: models/r21d/extract_r21d.py [hunk 2]
```diff
@@ -15,24 +15,33 @@
 PRE_CENTRAL_CROP_SIZE = (128, 171)
 KINETICS_MEAN = [0.43216, 0.394666, 0.37645]
 KINETICS_STD = [0.22803, 0.22145, 0.216989]
+R21D_MODEL_LIST = [
+    {'model_name': 'default', 'repo': None, 'stack_size': 16 , 'step_size': 16, 'dataset': 'kinetics'},
```
Comment (v-iashin) [inline models/r21d/extract_r21d.py:19]: If we change the `model_name` in the config, we also need to change it here.

File: models/r21d/extract_r21d.py [hunk 3]
```diff
@@ -15,24 +15,33 @@
 PRE_CENTRAL_CROP_SIZE = (128, 171)
 KINETICS_MEAN = [0.43216, 0.394666, 0.37645]
 KINETICS_STD = [0.22803, 0.22145, 0.216989]
+R21D_MODEL_LIST = [
```
Comment (v-iashin) [inline models/r21d/extract_r21d.py:18]: A bit of a nitpick but we could do it as a nested `dict` (`{<model_name>: {<param>: <value>}`. This would simplify the code in `__init__`.

```python
R21D_MODEL_DICT = {
    'r2plus1d_34_32_kinetics': {
        'repo': ...,
        'step_size': ...,   
    }
}
```

File: models/r21d/extract_r21d.py [hunk 4]
```diff
@@ -15,24 +15,33 @@
...
-DEFAULT_R21D_STACK_SIZE = 16
 
 class ExtractR21D(torch.nn.Module):
 
     def __init__(self, args):
         super(ExtractR21D, self).__init__()
         self.feature_type = args.feature_type
+        # select R21D model
+        valid_model_names = [m['model_name'] for m in R21D_MODEL_LIST]
+        assert args.model_name in valid_model_names, f'model_name: {args.model_name} not a valid model: {valid_model_names}'
```
Comment (v-iashin) [inline models/r21d/extract_r21d.py:34]: If `R21D_MODEL_DICT` is a dict, then this part becomes very brief and fluent:

```python
self.model_def = R21D_MODEL_DICT[args.model_name]
```
There would be no need for assert because it will fail with `KeyError` anyway.

File: models/r21d/extract_r21d.py [hunk 5]
```diff
@@ -133,7 +143,12 @@ def load_model(self, device: torch.device) -> Tuple[torch.nn.Module]:
         Returns:
             Tuple[torch.nn.Module]: the model with identity head, the original classifier
         '''
-        model = r2plus1d_18(pretrained=True)
+        model = torch.hub.load(
+                    self.model_def['repo'],
+                    model = self.model_def['model_name'],
+                    num_classes = self.model_def['num_classes'],
+                    pretrained = True
+                    ) if self.model_def['repo'] else r2plus1d_18(pretrained = True)
```
Comment (v-iashin) [inline models/r21d/extract_r21d.py:151]: It is great!

Can we make it more readable and explicit:

```python
# please check that the name is correct in config etc.
# please add `model_name` in `__init__`
if self.model_name == 'r2plus1d_18_16_kinetics':
    model = r2plus1d_18(pretrained=True)
else:
    model = torch.hub(....)
```

**Review by v-iashin — CHANGES_REQUESTED** [review]
Thanks a lot for the fixes. Here are a few nitpicks. I wrote them from the GitHub UI and not sure if my code does not have spelling errors and, hence, bugs. Please, follow your gut feeling.

**Comment by ohjho** [comment]
thanks for the review @v-iashin I agree to all of them and committed the necessary changes

**Review by v-iashin — APPROVED** [review]
I am going to create a new branch with these changes so I could test them during this week. Thanks a lot!
